### ISSUE TITLE
lightgbm categorical values from R

### ISSUE THREAD

**Issue by szilard** [szilard/GBM-perf#2]
@Laurae2

Is this the best way to deal with categorical variables in the lightgbm R package?

```
X_train_test <- sparse.model.matrix(dep_delayed_15min ~ .-1, data = rbind(d_train, d_test))

dlgb_train <- lgb.Dataset(data = X_train, label = ifelse(d_train$dep_delayed_15min=='Y',1,0))

md <- lgb.train(data = dlgb_train, objective = "binary",  ...
```

see full code [here](https://github.com/szilard/GBM-perf/blob/master/run/3-lightgbm.R).

It already gets awesome runtime and AUC, see [here](https://github.com/szilard/GBM-perf#gbm-performance).

**Comment by Laurae2** [comment]
@szilard All after point 1. are recommendations for comparison / lowering your AWS costs.

---

1. To deal with categorical variables, check this: https://github.com/Microsoft/LightGBM/pull/228 (LightGBM supports categorical features input, but in a special way).

If you are using one hot encoding, do not use one hot encoding and pass directly the categoricals as numeric.

Transforming a categorical to a numeric: `my_data$factor <- as.numeric(as.factor(my_data$factor))`. Train/Test should be joined first before transforming categoricals into numeric. Then, you are free to transform the data into a sparse matrix if you want comparable results with xgboost/h2o. After, pass the `categorical_feature` parameter with the named categorical columns to force LightGBM to consider these features as categorical.

aka: `model <- lgb.train(params, temp_train, 2, learning_rate=1, min_data = 1, min_hessian = 1, num_leaves = 1024, categorical_feature = c("job", "month"))` (see job & month will be considered categorical for training).

---

2. small recommended fix for speed

> ```r
> dlgb_train <- lgb.Dataset(data = X_train, label = ifelse(d_train$dep_delayed_15min=='Y',1,0))
> ```

Change to: `dlgb_train <- lgb.Dataset(data = X_train, label = as.numeric(d_train$dep_delayed_15min=='Y'))`, will be much faster

---

3. small recommended fix for getting comparable results vs xgboost/h2o

> ```r
> system.time({
> md <- lgb.train(data = dlgb_train, objective = "binary", 
>             nrounds = 100, num_leaves = 512, learning_rate = 0.1)
> })```

Change to:

```r
system.time({
md <- lgb.train(data = dlgb_train, objective = "binary", 
            nrounds = 100, max_depth = 10, num_leaves = 1024, learning_rate = 0.1)
})
```

---

4. load data faster:

> ```r
> library(readr)
> d_train <- read_csv("train-10m.csv")
> d_test <- read_csv("test.csv")
> ```

Change to:

```r
library(data.table)
d_train <- fread("train-10m.csv")
d_test <- fread("test.csv")
```

---

5. If you are interested only in AUC at the end, you can use this Wilcoxon approach which computes AUC much faster and can't possibly go faster in pure R (see https://github.com/Laurae2/R_benchmarking#area-under-the-curve-roc-benchmarks for details):

```r
function (preds, labels) {
    x1 = as.numeric(preds[labels == 1])
    n1 = as.numeric(length(x1))
    x2 = as.numeric(preds[labels == 0])
    n2 = as.numeric(length(x2))
    r = rank(c(x1, x2))
    return((sum(r[1:n1]) - n1 * (n1 + 1)/2)/(n1 * n2))
}
```

---

6. Seed must apply:

* Before the training;
* Without a `system.time` call as it executes extra functions (garbage collect), which changes the seed in an unpredictable way.

However, LightGBM is R seed invariant (not linking to R seed, only to an internal LightGBM seed) unlike xgboost.

---

7. Note about AWS:

* Cores provided are hyperthreaded cores (divide real performance by 2)
* Without a dedicated instance, results may vary widely

---

8. Getting more comparable results:

* Set loss sum (gamma) identical for each ML tool (xgboost/LightGBM can, no idea if h2o can)
* Set hessian sum pruning (hessian weights) identical for each ML tool (xgboost/LightGBM can, no idea if h2o can)
* You can test xgboost exact, fast histogram depthwise, and fast histogram lossguide (cannot compare xgboost exact vs LightGBM, it's like apples and oranges like h2o vs LightGBM => essentially, they may be GBDT, but they are way too different in reality)
* Set number of bins to an identical number, usually set to 255 because it makes each value hold in a 8-bit memory space (h2o / xgboost fast histogram / LightGBM) unless the ML tool is not optimized for that
* Get rid of all h2o background checks/computations during training, they are very expensive on large datasets (remove all h2o internal monitoring)

---

9. What's happening during training

Note that:

* LightGBM builds the dataset on 1st training (expensive pre-training)
* xgboost spreads data across threads on every training (expensive pre-training)
* h2o performs a massive amount of background checks/computations before/during training which should be disabled (expensive training)

---

10. Note about h2o.gbm: will be obsolete soon as they are switching to xgboost (they have PRs going on for xgboost integration in h2o).

**Comment by szilard** [comment]
Thanks @Laurae2 for awesome detailed suggestions. I'll go through them later on today (California time). Will post here updated timings. Thanks again.

**Comment by szilard** [comment]
Re **1** (best way to deal with cat. features):

Yeah, I was expecting there is a way to match in R what I've done previously on the command line
```
time ~/bin/LightGBM-2.0/lightgbm  data=train.libsvm  task=train  objective=binary \
     num_iterations=100  learning_rate=0.1  num_leaves=512  min_data_in_leaf=0 categorical_column=0,1,2,4,5,6
```
with the proper encoding kindly supplied by @guolinke [here](https://github.com/szilard/benchm-ml/issues/46#issuecomment-264624398).

I'm doing now [this](https://github.com/szilard/GBM-perf/blob/master/run/3-lightgbm2.R#L15-L23) and now it's even faster, see [here](https://github.com/szilard/GBM-perf/blob/master/results.txt#L95-L110).

What's weird though is that the previous way of doing it in R was matching better training time on the CL with as I've done it [here](https://github.com/szilard/GBM-perf/blob/master/run/old/3-lightgbm.txt#L51-L52).

E.g. for 10M records:
```
command line 47s
R one-hot sparse matrix 44s
R numeric encoding and pass categorical_features 37s
```

**Comment by szilard** [comment]
Re 2-10:

2 yeah

3 - see later

4 bad habit :)

5 I'm mainly interested in training time, but thanks

6,8,9 yeah, you can try to make the benchmark more precise, match the params etc. I was gaining interested in some quick results and if a tools is 5x better than the other, these things will not change much 

7 We both know of NUMA etc. maybe I should also run this while pinning the process to the cores on a single CPU/socket (and no hyperthreading)

10 I'm not sure, in some applications things like real-time scoring, easy deployment etc. matter more than just training time

---------------------

3: Sure, for 10M records:

```
lightgbm original   37s        0.7758
lightgbm new        28s        0.7563
xgboost.              290s        0.75110
```

I think I would trade 0.02 increase of AUC for just 30% increase in training time, which shows lightgbm's idea of trees of incomplete levels is really great.

Anyway, thanks again @Laurae2 for awesome help.

**Comment by guolinke** [comment]
@szilard @Laurae2 
Actually, you can also use one-hot coding for the categorical feature now. 
Its speed is almost the same as the ```categorical feature```.

**Comment by szilard** [comment]
@guolinke you mean directly on CSV file (command line version)?

**Comment by guolinke** [comment]
@szilard Both CSV or R's sparse matrix is okay. 

Also from your comment, I found you already test the ``R one-hot sparse matrix``, whose speed are almost the same as ```categorical_feature``` by numerical coding. This result is normal 😄 .

**Comment by szilard** [comment]
@guolinke Yeah, I did :)

@guolinke @Laurae2 Fantastic improvement in features since last I checked a few months ago. I'm really happy about the R package :)

With CSV from R I get:
```
> system.time({
+ md <- lightgbm(data = "train-1m.csv", objective = "binary",
+             nrounds = 100, num_leaves = 512, learning_rate = 0.1)
+ })
[LightGBM] [Fatal] Unknown token "c-12" in data file
Error in lgb.call("LGBM_DatasetCreateFromFile_R", ret = handle, lgb.c_str(private$raw_data),  :
  api error: Unknown token "c-12" in data file
```

The file is like:
```
"Month","DayofMonth","DayOfWeek","DepTime","UniqueCarrier","Origin","Dest","Distance","dep_delayed_15min"
"c-4","c-26","c-2",1828,"XE","LEX","IAH",828,"N"
"c-12","c-11","c-1",1212,"UA","DEN","MCI",533,"N"
```

**Comment by guolinke** [comment]
@szilard 
Oh, sorry, I didn't mean it can use the non-numerical data. 

There are only two ways for the categorical feature now:
1. convert to one-hot coding
2. map category to integer, and tell lightgbm they are ```categorical_feature``` .

My point in previous comment is, the speed of (1) and (2) are almost the same now.

**Comment by szilard** [comment]
Yeah, 1 and 2 above are almost same speed. 

I thought there is a 3rd way (directly strings), I get it now. I was confused that I've also seen in the Python example `panda.read_csv` https://github.com/Microsoft/LightGBM/blob/master/examples/python-guide/simple_example.py#L11 but I see now that file has only numeric data as well. Thanks @guolinke for clarification.

**Comment by guolinke** [comment]
@szilard 

I just open an new issue[https://github.com/Microsoft/LightGBM/issues/557] for the better categorical feature support in R.

**Comment by Laurae2** [comment]
@szilard Issue for better categorical support is fixed (*"smart converter"*), check https://github.com/Microsoft/LightGBM/pull/561 for details.

**Comment by szilard** [comment]
@Laurae2 great!!

**Comment by Laurae2** [comment]
@szilard Do you think a categorical preprocessor is appropriate so every model is using the same "dataset"? (catboost seems to numerically encode the categoricals)

**Comment by szilard** [comment]
I dunno :) It's really hard (impossible?) to match (exactly) the params and settings of the different algos. In this benchmarks and the original https://github.com/szilard/benchm-ml I've been just trying to see if there is a really large difference in algos (e.g. Spark is 10-100x slower, uses 10x more RAM and its accuracy is broken).

**Comment by guolinke** [comment]
@szilard 
I just update the categorical feature support in LightGBM (https://github.com/Microsoft/LightGBM/pull/762).

A simple test:

old solution:
```
[LightGBM] [Info] Trained a tree with leaves=255 and max_depth=28
[LightGBM] [Info] Iteration:100, valid_1 auc : 0.758738
[LightGBM] [Info] 46.584943 seconds elapsed, finished iteration 100
```

new solution:
```
[LightGBM] [Info] Trained a tree with leaves=255 and max_depth=19
[LightGBM] [Info] Iteration:100, valid_1 auc : 0.77983
[LightGBM] [Info] 85.287495 seconds elapsed, finished iteration 100
```
command: ```lightgbm.exe data=dataexpo.train test=dataexpo.test metric=auc app=binary categorical_column=0,1,2,4,5,6 num_leaves=255 learning_rate=0.1```

I think gap will be larger when running with more iterations.

**Comment by szilard** [comment]
@guolinke Sounds great, thanks for heads up. 

@Laurae2 @guolinke Will this be available in the R package as well?

**Comment by Laurae2** [comment]
@szilard The parameters should be exposed by default in the R wrapper I think.

**Comment by guolinke** [comment]
@szilard The optimal categorical feature support is temporary removed now. 
I will re-run the result when it is back.

**Comment by Laurae2** [comment]
New results on-demand with categorical support, example script with everything hardcoded:

```r
library(data.table)
library(lightgbm)
library(Matrix)
library(ROCR)

set.seed(123)

d_train <- fread("train-0.1m.csv", showProgress=FALSE, stringsAsFactors=TRUE)
d_test <- fread("test.csv", showProgress=FALSE, stringsAsFactors=FALSE)   ## to match factors in train and test with bind

d_train_test <- rbind(d_train, d_test)
n1 <- nrow(d_train)
n2 <- nrow(d_test)
d_train_test[, Month := as.numeric(Month)]
d_train_test[, DayofMonth := as.numeric(DayofMonth)]
d_train_test[, DayOfWeek := as.numeric(DayOfWeek)]
d_train_test[, UniqueCarrier := as.numeric(UniqueCarrier)]
d_train_test[, Origin := as.numeric(Origin)]
d_train_test[, Dest := as.numeric(Dest)]
X_train <- as.matrix(d_train_test[1:n1, 1:8])
X_test <- as.matrix(d_train_test[(n1+1):(n1+n2), 1:8])

dlgb_train <- lgb.Dataset(data = X_train, label = ifelse(d_train$dep_delayed_15min=="Y",1,0))


cat(system.time({
  md <- lgb.train(data = dlgb_train,
                  objective = "binary",
                  nrounds = 100, num_leaves = 512, learning_rate = 0.1,
                  device = "cpu", nthread = 18, categorical_feature = c(1, 2, 3, 5, 6, 7),
                  verbose = 0)
})[[3]]," ",sep="")

phat <- predict(md, data = X_test)
rocr_pred <- prediction(phat, d_test$dep_delayed_15min)
cat(performance(rocr_pred, "auc")@y.values[[1]],"\n")
```

Copied hardware from here: https://github.com/szilard/GBM-perf/issues/12

CPU only with 18 physical threads (`numactl` for 1st socket + OpenMP environment variables lock in):

| ?gb | Size | Speed (s) | AUC |
| ---: | ---: | ---: | ---: |
| xgb ohe | 0.1m | 4.181 | 0.7324224 |
| xgb ohe | 1m | 15.978 | 0.7494959 |
| xgb ohe | 10m | 104.598 | 0.7551197 |
| lgb ohe | 0.1m | 1.763 | 0.7298355 |
| lgb ohe | 1m | 4.253 | 0.7636987 |
| lgb ohe | 10m | 38.197 | 0.7742033 |
| lgb cat | 0.1m | 1.855 | 0.7170897 |
| lgb cat | 1m | 5.559 | 0.7649034 |
| lgb cat | 10m | 54.758 | 0.7918742 |

N.B: as LightGBM parallelism is by number of columns, then the number of threads is set to 8 for categorical handling. There are other parts of LightGBM which can benefit from more threads but they should be negligible due to the size of data.

**Comment by szilard** [comment]
Thanks @Laurae2 for updated results.

**Comment by szilard** [comment]
Forgot about this discussion and I opened recently a new issue on this topic:

https://github.com/szilard/GBM-perf/issues/38

Also see there that I changed the benchmark to use cat.enc. instead of OHE, so I think we can close this one.
